01/highest-rated.md

```diff
@@ -0,0 +1,183 @@
+
+1.  What should the function return if there are multiple restaurants tied for the highest rating?
+1.  How should the function handle restaurant dictionaries with missing data (name or rating)?
+1.  How should the function handle non-numerical values for `"rating"`?
+1. 
+1.
+
+
+
+##### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 2 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 890d7045-28cb-4832-812b-d65803ce2618
+* title: Write unit tests
+* topics: pse
+##### !question
+
+Write at least two unit tests for `get_highest_rated`. 
+* Consider at least one nominal and one edge test case.
+* What is the expected output for the given input?
+
+*Note: Click the **Run Tests** button to save your tests for instructor feedback. No real tests are actually run again your unit tests.*
+##### !end-question
+
+##### !placeholder
+
+```py
+def test_nominal_case():
+    # ^rename with meaninful test name
+    # and complete test implementation below
+    pass
+    # arrange
+
+    # act
+
+    # assert
+
+def test_edge_case():
+    pass
+```
+
+##### !end-placeholder
+
+##### !tests
+
+```py
+import unittest
+
+class TestPython1(unittest.TestCase):
+  def test_always_pass(self):
+    self.assertEqual(1,1)
+```
+
+##### !end-tests
+##### !explanation 
+
+Example tests:
+
+```python
+def test_picks_highest_rated_from_list():
+    # arrange
+    restaurants = [
+            {"name": "Grillby's", "rating": 1},
+            {"name": "Crow's Nest", "rating": 5}
+        ]
+
+    # act
+    output = get_highest_rated(restaurants)
+    
+    # assert
+    assert output["name"] == "Crow's Nest"
+    assert output["rating"] == 5
+    assert len(output.keys()) == 2
+
+def test_picks_from_list_of_one():
+    # arrange
+    restaurants = [{"name": "Crow's Nest", "rating": 1}]
+
+    # act
+    output = get_highest_rated(restaurants)
+    
+    # assert
+    assert output["name"] == "Crow's Nest"
+    assert output["rating"] == 1
+    assert len(output.keys()) == 2
+
+def test_returns_none_with_zero_restaurants(self):
+    # arrange
+    restaurants = []
+
+    # act
+    output = get_highest_rated(restaurants)
+    
+    # assert
+    assert output == None
+```
+
+##### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: paragraph
+* id: 085ad8a1-7dbd-49ad-aa7a-2213dd0d3c76
+* title: Create Logical Steps
+* topics: pse
+
+##### !question
+
+Without writing code, describe how you would implement `get_highest_rating` in enough detail that someone else could write the code. 
+* Your logical steps could take the form of a numbered list, pseudo code, or anywhere inbetween. What's important at this stage is to think through and outline the implementation before writing code.
```

Comment:
Yes! At first I had a bullet point here, but I was having trouble getting away from the word "subproblem". I was considering something like.

* Consider whether it would be helpful to first break the problem up into subproblem. For example, 1. Validate input, 2. Given valid input, perform the computation/solve the problem/etc.

What do you think?

Reply:
I like that suggestion!